p2panda-rs/src/permissions/key_group/key_group.rs

```diff
@@ -0,0 +1,297 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::collections::HashMap;
+use std::convert::TryFrom;
+
+use log::debug;
+
+use crate::document::{Document, DocumentId, DocumentViewId};
+use crate::identity::Author;
+use crate::operation::{Operation, OperationFields, OperationValue, PinnedRelation, Relation};
+use crate::permissions::key_group::{
+    KeyGroupError, KeyGroupView, Membership, MembershipRequestView, MembershipResponseView, Owner,
+};
+use crate::schema::SchemaId;
+
+/// Represents a group of key pairs that can be assigned shared ownership of documents.
+#[derive(Debug, Clone)]
+pub struct KeyGroup {
+    id: DocumentId,
+    view_id: DocumentViewId,
+    name: String,
+    members: HashMap<Author, Vec<Membership>>,
+}
+
+impl KeyGroup {
+    /// Create a key group from a key group view and a set of memberships.
+    ///
+    /// The members parameter must only contain memberships of the key group to be created.
+    /// The `member_key_groups` parameter must contain all key groups that have memberships and may
+    /// contain additional unrelated key groups.
+    pub fn new(
+        key_group: &KeyGroupView,
+        members: &[Membership],
+        member_key_groups: &[KeyGroup],
+    ) -> Result<KeyGroup, KeyGroupError> {
+        debug!("Building {}", key_group.name());
+
+        // Collect all (author, membership) pairs from `members` parameter, including duplicate
+        // author values.
+        let mut member_pool = vec![];
+        for membership in members {
+            match membership.member() {
+                // Simple case: for single key memberships just add that key to the pool.
+                Owner::Author(value) => {
+                    debug!("Adding author {:?}", value.as_str());
+                    member_pool.push((value, membership));
+                }
+
+                // When a key group is a member, recursively add that key group's members to the
+                // pool, assigned to a shared `membership`
```

Comment:
Does this account for when a key group is a member of a key group which is a member of a key group? (you know what I mean ;-p)